### PR TITLE
Fixes #34975 - Removed ks=. is supplied by kernel options snippet

### PR DESCRIPTION
**Pull request by ShuffleBox**
Fixes [#34975](https://projects.theforeman.org/issues/34975)

Issue encountered with Satellite 6.10.5.1 and RedHat Enterprise Linux 9.0.  RHEL9 does not begin installation as ks= kernel boot parameter is being passed instead of inst.ks=.

This bug is similar to the work done on https://projects.theforeman.org/issues/32486 the kickstart_default_ipxe template renders the following currently.

```
#!gpxe


echo Trying to ping Gateway: ${netX/gateway}
ping --count 1 ${netX/gateway} || echo Ping to Gateway failed or ping command not available.
echo Trying to ping DNS: ${netX/dns}
ping --count 1 ${netX/dns} || echo Ping to DNS failed or ping command not available.

kernel http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/vmlinuz initrd=initrd.img ks=http://satellite.fqdn:8000/unattended/provision?token=d79a2f7d-cfdc-49cc-ac81-88f9761d572c  BOOTIF=01-00-50-56-ac-10-13 kssendmac ks.sendmac inst.ks.sendmac ip=dhcp
initrd http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/initrd.img
imgstat
sleep 2
boot
```

By removing `ks=<%= ks %>` the desired `inst.ks=` segment is rendered from the kickstart_kernel_options snippet.
```
#!gpxe


echo Trying to ping Gateway: ${netX/gateway}
ping --count 1 ${netX/gateway} || echo Ping to Gateway failed or ping command not available.
echo Trying to ping DNS: ${netX/dns}
ping --count 1 ${netX/dns} || echo Ping to DNS failed or ping command not available.

kernel http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/vmlinuz initrd=initrd.img  BOOTIF=01-00-50-56-ac-10-13 inst.ks=http://satellite.fqdn:8000/unattended/provision?token=d79a2f7d-cfdc-49cc-ac81-88f9761d572c inst.ks.sendmac ip=dhcp
initrd http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/initrd.img
imgstat
sleep 2
boot
```

Side note: I attempted to regenerate the tests with `RAILS_ENV=test bundle exec rake snapshots:generate`, but there may be an error in a CoreOS template that is causing it to fail half way through.

edit: fixed copy/paste problem for accurate description of the request.

### CODE REVIEW

**Comment by adamruzicka** [comment]
The changes proposed here look good, however could you please rebase so we get rid of the `Merge branch 'theforeman:develop' into develop` commit and reword the commit message of the actual change to be in the form `Fixes #34975 - $description`?

**Comment by adamruzicka** [comment]
The changes proposed here look good, however could you please rebase so we get rid of the `Merge branch 'theforeman:develop' into develop` commit and reword the commit message of the actual change to be in the form `Fixes #34975 - $description`?

**Comment by ShuffleBox** [comment]
> The changes proposed here look good, however could you please rebase so we get rid of the `Merge branch 'theforeman:develop' into develop` commit and reword the commit message of the actual change to be in the form `Fixes #34975 - $description`?

I stumbled a bit with git, but I think I finally got everything where it needs to be.  Please let me know if anything needs to be revisited.

**Comment by adamruzicka** [comment]
Looks good to me.

@ofedoren I vaguely recall you mentioning you wanted to check something in here? Or maybe I'm misremembering

File: app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb
Comment (ofedoren) [inline app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb:43]: If we remove this entry, do we still need https://github.com/theforeman/foreman/blob/develop/app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb#L25-L35 block?

Also, I'm not a provisioning guy, but what's the difference between https://github.com/theforeman/foreman/blob/0776ee953497a9772a4ca1bd86c68d9c5cf9051a/app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb#L30 and https://github.com/theforeman/foreman/blob/0776ee953497a9772a4ca1bd86c68d9c5cf9051a/app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb#L39 ?

**Review by ofedoren — COMMENTED** [review]
Thanks, @ShuffleBox, I've got only two questions.

Also, it seems like the actual tests didn't run...

**Comment by ofedoren** [comment]
ok to test

File: app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb
Comment (ShuffleBox) [reply app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb:43]: > If we remove this entry, do we still need https://github.com/theforeman/foreman/blob/develop/app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb#L25-L35 block?
> 
> Also, I'm not a provisioning guy, but what's the difference between
> 
> https://github.com/theforeman/foreman/blob/0776ee953497a9772a4ca1bd86c68d9c5cf9051a/app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb#L30
> and
> 
> https://github.com/theforeman/foreman/blob/0776ee953497a9772a4ca1bd86c68d9c5cf9051a/app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb#L39
> ?

I'll preface this with admitting to be a fairly infrequent contributor, and not an expert in Ruby templates. This may go more under the hood with foreman_url() and various kickstart deployment scenarios than I am able to speak on.  :) 

RE: kickstart_default_ipxe.erb#L25-L35
That being said, I believe you're right.  kickstart_default_ipxe.erb Line 29 - 36 looks to be setting the value of ks and is making determinations based on if it has a static IPv4/IPv6 address set or is DHCP.  As far as I know, foreman_url('provisioning', static: '1') would render the foreman url with the static=1 url parameter.  I haven't seen this on ks= urls in my deployment, but I do not have Satellite integrating with DNS & DHCP services where maybe this comes into play?

I think the kickstart_default_ipxe.erb L25-35 block could be integrated with the kickstart_kernel_options L36-40 If/Else block, if needed, and should arrive at a workable result.

Comment (ofedoren) [reply app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb:43]: Hmm, I agree with your proposal since it would allow us to be safer with the changes. Thus said, let's move `ks` variable setting from `kickstart_default_ipxe` into `kickstart_kernel_options` snippet. This way we don't actually remove the logic and we wouldn't set `ks` variable up more than once (as I see this PR is about). Could you please update?

Comment (ShuffleBox) [reply app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb:43]: Will do!  I couldn't get this today, but will try to get it in tomorrow.

Comment (ShuffleBox) [reply app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb:43]: I think I got it sorted.  Please review the latest commit.

I even found how to exercise this by configuring Static/DHCP on the Boot Mode option under Infrastructure -> Subnets.

Static IPV4:
```
#!gpxe


echo Trying to ping Gateway: ${netX/gateway}
ping --count 1 ${netX/gateway} || echo Ping to Gateway failed or ping command not available.
echo Trying to ping DNS: ${netX/dns}
ping --count 1 ${netX/dns} || echo Ping to DNS failed or ping command not available.

kernel http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/vmlinuz initrd=initrd.img  BOOTIF=01-00-50-56-ac-10-13 inst.ks=http://satellite.fqdn:8000/unattended/provision?static=1&token=7a2f900e-7ce7-4b61-885b-fd0d15d4872f ip=${netX/ip}::${netX/gateway}:${netX/netmask}:provisionhost.fqdn:ens192:none nameserver=${dns}
initrd http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/initrd.img
imgstat
sleep 2
boot

```

DHCP:
```
#!gpxe


echo Trying to ping Gateway: ${netX/gateway}
ping --count 1 ${netX/gateway} || echo Ping to Gateway failed or ping command not available.
echo Trying to ping DNS: ${netX/dns}
ping --count 1 ${netX/dns} || echo Ping to DNS failed or ping command not available.

kernel http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/vmlinuz initrd=initrd.img  BOOTIF=01-00-50-56-ac-10-13 inst.ks=http://satellite.fqdn:8000/unattended/provision?token=7a2f900e-7ce7-4b61-885b-fd0d15d4872f inst.ks.sendmac ip=dhcp
initrd http://satellite.fqdn/pulp/repos/Default_Organization/Library/Red_Hat_Enterprise_Linux_9_Kickstart/content/dist/rhel9/9.0/x86_64/baseos/kickstart//images/pxeboot/initrd.img
imgstat
sleep 2
boot
```

**Comment by ofedoren** [comment]
Thanks, @ShuffleBox, sorry for not revisiting in time, but could you also update the snapshots? This should fix the tests and actually show diffs between old rendered templates and the new ones.

You can do it by running `bundle exec rake snapshots:generate` from the checkout.

**Comment by ShuffleBox** [comment]
> 

I ran `bundle exec rake snapshots:generate`.  However, it gave feedback that this can only be run in test mode.  I then ran `RAILS_ENV=test bundle exec rake snapshots:generate`.  This wrote some new files, but got hung up on "coreos_provision.erb" and did not complete, as far as I can tell.

```
[garret@foreman8 foreman]$ RAILS_ENV=test bundle exec rake snapshots:generate
Writing /home/garret/foreman/test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Alterator_default.host4dhcp.snap.txt
<snip - more successfully written snapshots>
Writing /home/garret/foreman/test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/AutoYaST_default_user_data.host4dhcp.snap.txt
Writing /home/garret/foreman/test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/UserData_open-vm-tools.host4dhcp.snap.txt
rake aborted!
TypeError: no implicit conversion of false into String
/home/garret/foreman/lib/foreman/command_runner.rb:15:in `initialize'
/home/garret/foreman/lib/foreman/command_runner.rb:15:in `new'
/home/garret/foreman/lib/foreman/command_runner.rb:15:in `initialize'
/home/garret/foreman/app/services/foreman/renderer/scope/macros/transpilers.rb:18:in `new'
/home/garret/foreman/app/services/foreman/renderer/scope/macros/transpilers.rb:18:in `transpile_coreos_linux_config'
/home/garret/foreman/vendor/ruby/2.7.0/gems/safemode-1.3.7/lib/safemode/scope.rb:37:in `method_missing'
CoreOS provision:73:in `bind'
/home/garret/foreman/vendor/ruby/2.7.0/gems/safemode-1.3.7/lib/safemode.rb:51:in `eval'
/home/garret/foreman/vendor/ruby/2.7.0/gems/safemode-1.3.7/lib/safemode.rb:51:in `eval'
/home/garret/foreman/app/services/foreman/renderer/safe_mode_renderer.rb:7:in `render'
/home/garret/foreman/app/services/foreman/renderer/base_renderer.rb:18:in `render'
/home/garret/foreman/app/services/foreman/renderer.rb:46:in `render'
/home/garret/foreman/app/services/foreman/template_snapshot_service.rb:45:in `render_template'
/home/garret/foreman/lib/tasks/snapshots.rake:40:in `block (6 levels) in <top (required)>'
/home/garret/foreman/lib/tasks/snapshots.rake:35:in `each'
/home/garret/foreman/lib/tasks/snapshots.rake:35:in `block (5 levels) in <top (required)>'
/home/garret/foreman/lib/tasks/snapshots.rake:34:in `each'
/home/garret/foreman/lib/tasks/snapshots.rake:34:in `block (4 levels) in <top (required)>'
/home/garret/foreman/app/models/concerns/foreman/thread_session.rb:108:in `as'
/home/garret/foreman/lib/tasks/snapshots.rake:33:in `block (3 levels) in <top (required)>'
/home/garret/foreman/vendor/ruby/2.7.0/gems/database_cleaner-1.99.0/lib/database_cleaner/generic/base.rb:16:in `cleaning'
/home/garret/foreman/vendor/ruby/2.7.0/gems/forwardable-1.3.2/lib/forwardable.rb:238:in `cleaning'
/home/garret/foreman/vendor/ruby/2.7.0/gems/database_cleaner-1.99.0/lib/database_cleaner/configuration.rb:88:in `block (2 levels) in cleaning'
/home/garret/foreman/vendor/ruby/2.7.0/gems/database_cleaner-1.99.0/lib/database_cleaner/configuration.rb:89:in `cleaning'
/home/garret/foreman/vendor/ruby/2.7.0/gems/forwardable-1.3.2/lib/forwardable.rb:238:in `cleaning'
/home/garret/foreman/lib/tasks/snapshots.rake:25:in `block (2 levels) in <top (required)>'
/home/garret/foreman/vendor/ruby/2.7.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `load'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `kernel_load'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:28:in `run'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:476:in `exec'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'
/usr/share/gems/gems/bundler-2.1.4/libexec/bundle:46:in `block in <top (required)>'
/usr/share/gems/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
/usr/share/gems/gems/bundler-2.1.4/libexec/bundle:34:in `<top (required)>'
/usr/bin/bundle:23:in `load'
/usr/bin/bundle:23:in `<main>'
Tasks: TOP => snapshots:generate
(See full trace by running task with --trace)
```

This also seemed to occur with a fresh clone of theforeman:develop.

**Comment by ShuffleBox** [comment]
It appears that the snapshot generation issue was reported a few months ago.  [Bug #34664 -generate snapshots fails because of transpile_coreos_linux_config](https://projects.theforeman.org/issues/34664)

**Comment by ofedoren** [comment]
Thanks, @ShuffleBox, I've updated the snapshots for you :) Although there are snapshots that I though wouldn't be touched, but it seems like we are fixing even a different bug? Not sure though. Otherwise looks good. Let's wait until it's :green_apple: and then I'll merge.

UPD: I'll merge, but only after @adamruzicka approval ;)

UPD2: Test failures don't seem to be related. They just fail in the same way for any other PR :/

UPD3: https://github.com/theforeman/foreman/pull/9288

**Comment by ShuffleBox** [comment]
@ofedoren  Thank you for the help with the snapshots.  For my own edification, is the failure I experienced generating snapshots a bug, or more likely something wrong with my environment?

**Review by ofedoren — APPROVED** [review]
Thanks, @ShuffleBox, let's get this in :green_apple:

**Comment by ShuffleBox** [comment]
Thank you for the opportunity to participate!

**Comment by ofedoren** [comment]
CP 3.3: https://github.com/theforeman/foreman/pull/9308

**Pull request merged**
